File: browser/src/js/moneysocket/nexus/consumer.js
// Copyright (c) 2020 Jarret Dyrbye
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php

const Timestamp = require('../utl/timestamp.js').Timestamp;
const Nexus = require("./nexus.js").Nexus;

const RequestProvider = require(
    "../message/request/provider.js").RequestProvider;
const RequestPing = require("../message/request/ping.js").RequestPing;


const LAYER_NOTIFICATIONS = new Set(["NOTIFY_PROVIDER",
                                     "NOTIFY_PROVIDER_NOT_READY",
                                     "NOTIFY_PONG",
                                   ]);

class ConsumerNexus extends Nexus {
    constructor(below_nexus, layer) {
        super(below_nexus, layer);

        this.onproviderinfo = null;
        this.onping = null;

        this.consumerFinishedCb = null;
        this.handshake_finished = false;
        this.ping_interval = null;
        this.ping_start_time = null;
    }

    ///////////////////////////////////////////////////////////////////////////

    isLayerMessage(msg) {
        if (msg['message_class'] != "NOTIFICATION") {
            return false;
        }
        return LAYER_NOTIFICATIONS.has(msg['notification_name']);
    }

    onMessage(below_nexus, msg) {
        //console.log("consumer nexus got message");
        if (! this.isLayerMessage(msg)) {
            super.onMessage(below_nexus, msg)
            return;
        }
        if (msg['notification_name'] == "NOTIFY_PROVIDER") {
            if (! this.handshake_finished) {
                this.handshake_finished = true;
                this.consumerFinishedCb(this);
            }
            if (this.onproviderinfo != null) {
                this.onproviderinfo(this, msg);
            }
        } else if (msg['notification_name'] == "NOTIFY_PROVIDER_NOT_READY") {
            console.log("provider not ready, waiting");
        } else if (msg['notification_name'] == "NOTIFY_PONG") {
            var msecs = (Timestamp.getNowTimestamp() -
                         this.ping_start_time) * 1000;
            if (this.onping != null) {
                this.onping(this, Math.round(msecs));
            }
            this.ping_start_time = null;
        }
    }

    onBinMessage(below_nexus, msg_bytes) {
        //console.log("consumer nexus got raw msg from below");
    }

    startHandshake(consumerFinishedCb) {
        this.consumerFinishedCb = consumerFinishedCb;
        this.send(new RequestProvider());
    }

    startPinging() {
        console.log("START PING");
        this.ping_interval = setInterval(
            function() {
                this.ping_start_time = Timestamp.getNowTimestamp();
                this.send(new RequestPing());
            }.bind(this), 3000);
        this.ping_start_time = Timestamp.getNowTimestamp();
        this.send(new RequestPing());
    }

    stopPinging() {
        console.log("STOP PING");
        clearInterval(this.ping_interval);
        this.ping_interval = null;
    }
}

exports.ConsumerNexus = ConsumerNexus;

File: python/moneysocket/nexus/consumer.py
# Copyright (c) 2020 Jarret Dyrbye
# Distributed under the MIT software license, see the accompanying
# file LICENSE or http://www.opensource.org/licenses/mit-license.php

import uuid
import time
import logging

from twisted.internet.task import LoopingCall

from moneysocket.nexus.nexus import Nexus

from moneysocket.message.request.provider import RequestProvider
from moneysocket.message.request.ping import RequestPing


class ConsumerNexus(Nexus):
    def __init__(self, below_nexus, layer):
        super().__init__(below_nexus, layer)
        self.onping = None
        self.consumer_finished_cb = None;
        self.handshake_finished = False;
        self.ping_interval = None;
        self.ping_start_time = None;


    ###########################################################################

    def is_layer_message(self, msg):
        if msg['message_class'] != "NOTIFICATION":
            return False
        return msg['notification_name'] in {"NOTIFY_PROVIDER",
                                            "NOTIFY_PROVIDER_NOT_READY",
                                            "NOTIFY_PING"}

    def on_message(self, below_nexus, msg):
        logging.info("consumer nexus got msg")

        if not self.is_layer_message(msg):
            super().on_message(below_nexus, msg)
            return

        if msg['notification_name'] == "NOTIFY_PROVIDER":
            if not self.handshake_finished:
                self.handshake_finished = True
                # nexus is announced to the transact layer
                # TODO, this message might be better to handle in the transact
                # layer instead. Refactor might be needed.
                self.consumer_finished_cb(self)
            # pass the message above for the transact layer to process
            super().on_message(below_nexus, msg)
        elif msg['notification_name'] == "NOTIFY_PROVIDER_NOT_READY":
            logging.info("provider not ready, waiting")
        elif msg['notification_name'] == "NOTIFY_PONG":
            if not self.ping_start_time:
                return
            msecs = (time.time() - self.ping_start_time) * 1000;
            if self.onping:
                self.onping(self, round(msecs))
            self.ping_start_time = None;

    def on_bin_message(self, below_nexus, msg_bytes):
        logging.info("provider nexus got raw msg")
        super().on_bin_message(below_nexus, msg_bytes)

    ###########################################################################


    def start_handshake(self, consumer_finished_cb):
        self.consumer_finished_cb = consumer_finished_cb;
        self.send(RequestProvider());

    def send_ping(self):
        self.ping_start_time = time.time()
        self.send(RequestPing());


    def start_pinging(self):
        logging.info("START PING");
        self.ping_interval = LoopingCall(self.send_ping)
        self.ping_interval.start(3.0, now=True)

    def stop_pinging(self):
        logging.info("STOP PING");
        if not self.ping_interval:
            return
        self.ping_interval.stop()
        self.ping_interval = None;
